### PR TITLE
enable ionization for CPU backend

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- remove define which disabled the ionization code on CPU
- rewrite kernel with the lockstep programming model
- add documentation
- use new functor interface with `collectiveInit`
- use a source frame for each worker (instead of all worker in the block)
- extend interface of all functors used together with `CreateParticlesKernel` (add method `collectiveInit()`)

# Review

Please check the changes commit wise!

# Tests

- [x] LWFA test case 10 (CPU/GPU vs GPU 0.3.X)
- [x] rebase against #2233

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
ionizationcompare

The number of particles spawned on the cpu in peak ~5000 particles higher than on the gpu (release 0.3.0).
The new alpaka cuda version is in peak by a factor 1.13 slower than the release 0.3.0, this is coming from #2226 where the minimal state of the random number generator is removed.

**Comment by ax3l** [comment]
code looks good!
the test would be good to be repeated with the check for free charge in the `FoilLCT` since the Hydrogen ionization physics in the LWFA example (with a strong `a_0`) is rather insensitive to potential bugs.

File: include/picongpu/particles/creation/creation.kernel [hunk 1]
```diff
@@ -30,282 +30,458 @@
...
+                        )
+                        {
+                            if( 0 <= targetParIdCtx[ idx ] &&  targetParIdCtx[ idx ] < maxParticlesInFrame )
+                            {
+                                // each thread makes the attributes of its source particle accessible
+                                auto sourceParticle = sourceFrame[ linearIdx ];
+                                // each thread initializes an target particle if one should be created
+                                auto targetParticle = targetFrame[ targetParIdCtx[ idx ] ];
+
+                                // create an target particle in the new target particle frame:
```
Comment (n01r) [inline include/picongpu/particles/creation/creation.kernel]: ... `a target` ...

File: include/picongpu/particles/creation/creation.kernel [hunk 2]
```diff
@@ -30,282 +30,458 @@
...
+                            uint32_t const idx
+                        )
+                        {
+                            if( targetParIdCtx[ idx ] >= maxParticlesInFrame )
+                            {
+                                targetParIdCtx[ idx ] -= maxParticlesInFrame;
+
+                                // each thread makes the attributes of its source particle accessible
+                                auto sourceParticle = sourceFrame[ linearIdx ];
+                                // each thread initializes an target particle if one should be created
```
Comment (n01r) [inline include/picongpu/particles/creation/creation.kernel]: ... `a target` ...

File: include/picongpu/particles/creation/creation.kernel [hunk 3]
```diff
@@ -30,282 +30,458 @@
...
+                                auto targetParticle = targetFrame[ targetParIdCtx[ idx ] ];
+
+                                // create an target particle in the new target particle frame:
```
Comment (n01r) [inline include/picongpu/particles/creation/creation.kernel]: ... `a target` ...

File: include/picongpu/particles/bremsstrahlung/Bremsstrahlung.tpp [hunk 4]
```diff
@@ -76,13 +77,14 @@ template<
     typename T_ElectronSpecies,
     typename T_PhotonSpecies
 >
-template< typename T_Acc >
-DINLINE
-void Bremsstrahlung<T_IonSpecies, T_ElectronSpecies, T_PhotonSpecies>::init(
-    T_Acc const & acc,
+template<
+    typename T_Acc ,
```
Comment (n01r) [inline include/picongpu/particles/bremsstrahlung/Bremsstrahlung.tpp]: The docstrings are missing here as well but the `T_Acc acc` is probably in every kernel, right? So the question arises if it makes sense to document its every occurence. I'd prefer it personally, though.

**Comment by n01r** [comment]
> ... the test would be good to be repeated with the check for free charge in the FoilLCT ...

I agree with @ax3l on the testing. A hydrogen-only setup will not be prone to many potential mistakes as there is just one ionization process per ion.

File: include/picongpu/particles/bremsstrahlung/Bremsstrahlung.tpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/picongpu/particles/bremsstrahlung/Bremsstrahlung.tpp]: the documentation is in the `hpp` file

Comment (n01r) [reply include/picongpu/particles/bremsstrahlung/Bremsstrahlung.tpp]: ah, alright, sorry :)

**Comment by psychocoderHPC** [comment]
# Test

script to create the plots will be provided by @ax3l

ionization

File: include/picongpu/particles/creation/creation.kernel [hunk 5]
```diff
@@ -30,282 +30,458 @@
...
+            sourceBox( sourceBox ),
+            targetBox( targetBox ),
+            particleCreator( particleCreator ),
+            guardSuperCells( guardSuperCells )
+        { }
+
+        /** Goes over all frames and calls `ParticleCreator`
+         *
+         * @param blockCell n-dim. block offset (in cells) relative to the origin
+         *                  of the local domain plus guarding cells
```
[the remark this answers is not in the document]
Comment (ax3l) [reply include/picongpu/particles/creation/creation.kernel]: added

**Pull request merged**